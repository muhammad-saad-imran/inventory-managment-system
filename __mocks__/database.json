{
  "clients": [
    {
      "id": "e4b6a0f1-d09e-4e95-8d1a-8e091b2a7510",
      "name": "Alice Smith",
      "email": "alice.smith@example.com",
      "phone_number": 1234567890,
      "created_at": "2024-01-01T12:00:00Z"
    },
    {
      "id": "c3f7d832-04e1-4b77-b19b-5a7c36be83d1",
      "name": "Bob Johnson",
      "email": "bob.johnson@example.com",
      "phone_number": 9876543210,
      "created_at": "2024-01-05T15:30:00Z"
    }
  ],
  "products": [
    {
      "id": "f2b2d5eb-76d4-4b09-a64d-9c2fd6a4d19e",
      "name": "Product A",
      "description": "Description of Product A",
      "created_at": "2024-02-01T09:00:00Z"
    },
    {
      "id": "a4cde3f2-1f98-4b53-b8ee-8f7d64a84b9d",
      "name": "Product B",
      "description": "Description of Product B",
      "created_at": "2024-02-15T10:00:00Z"
    }
  ],
  "suppliers": [
    {
      "id": "b9d6d220-4f6c-42d8-9083-b0a5c3f05a9a",
      "name": "Supplier One",
      "email": "supplier.one@example.com",
      "phone_number": 5551234567,
      "created_at": "2024-03-01T08:00:00Z"
    },
    {
      "id": "a2b5d5b9-4c3a-4b51-a2c6-123456789abc",
      "name": "Supplier Two",
      "email": "supplier.two@example.com",
      "phone_number": 5559876543,
      "created_at": "2024-03-10T14:00:00Z"
    }
  ],
  "inventory": [
    {
      "id": "edc1b083-5c7f-4d7f-8d52-1c5302a3d7b1",
      "cost": 50.00,
      "supplier_id": "b9d6d220-4f6c-42d8-9083-b0a5c3f05a9a",
      "product_id": "f2b2d5eb-76d4-4b09-a64d-9c2fd6a4d19e",
      "stock_quantity": 100,
      "reorder_limit": 20,
      "supply_date": "2024-03-15T10:00:00Z",
      "created_at": "2024-03-15T10:00:00Z",
      "suppliers": {
        "id": "b9d6d220-4f6c-42d8-9083-b0a5c3f05a9a",
        "name": "Supplier One",
        "email": "supplier.one@example.com",
        "phone_number": 5551234567,
        "created_at": "2024-03-01T08:00:00Z"
      },
      "products": {
        "id": "f2b2d5eb-76d4-4b09-a64d-9c2fd6a4d19e",
        "name": "Product A",
        "description": "Description of Product A",
        "created_at": "2024-02-01T09:00:00Z"
      }
    },
    {
      "id": "ac15d5a6-14f1-4a7f-98a3-0e4f2e1c1d4f",
      "cost": 75.00,
      "supplier_id": "a2b5d5b9-4c3a-4b51-a2c6-123456789abc",
      "product_id": "a4cde3f2-1f98-4b53-b8ee-8f7d64a84b9d",
      "stock_quantity": 50,
      "reorder_limit": 10,
      "supply_date": "2024-03-20T11:00:00Z",
      "created_at": "2024-03-20T11:00:00Z",
      "suppliers": {
        "id": "a2b5d5b9-4c3a-4b51-a2c6-123456789abc",
        "name": "Supplier Two",
        "email": "supplier.two@example.com",
        "phone_number": 5559876543,
        "created_at": "2024-03-10T14:00:00Z"
      },
      "products": {
        "id": "a4cde3f2-1f98-4b53-b8ee-8f7d64a84b9d",
        "name": "Product B",
        "description": "Description of Product B",
        "created_at": "2024-02-15T10:00:00Z"
      }
    }
  ],
  "orders": [
    {
      "id": "e4b6a0f1-d09e-4e95-8d1a-8e091b2a7511",
      "client_id": "e4b6a0f1-d09e-4e95-8d1a-8e091b2a7510",
      "order_date": "2024-04-01T12:00:00Z",
      "status": "PROCESSING",
      "created_at": "2024-04-01T12:00:00Z",
      "clients": {
        "id": "e4b6a0f1-d09e-4e95-8d1a-8e091b2a7510",
        "name": "Alice Smith",
        "email": "alice.smith@example.com",
        "phone_number": 1234567890,
        "created_at": "2024-01-01T12:00:00Z"
      },
      "order_items": [
        {
          "id": "d8e10f0c-bb27-4d8d-bff6-46f9e925a5b1",
          "order_id": "e4b6a0f1-d09e-4e95-8d1a-8e091b2a7511",
          "inventory_id": "edc1b083-5c7f-4d7f-8d52-1c5302a3d7b1",
          "quantity": 2,
          "price": 100.00,
          "created_at": "2024-04-01T12:00:00Z",
          "inventory": {
            "id": "edc1b083-5c7f-4d7f-8d52-1c5302a3d7b1",
            "cost": 50.00,
            "supplier_id": "b9d6d220-4f6c-42d8-9083-b0a5c3f05a9a",
            "product_id": "f2b2d5eb-76d4-4b09-a64d-9c2fd6a4d19e",
            "stock_quantity": 100,
            "reorder_limit": 20,
            "supply_date": "2024-03-15T10:00:00Z",
            "created_at": "2024-03-15T10:00:00Z",
            "suppliers": {
              "id": "b9d6d220-4f6c-42d8-9083-b0a5c3f05a9a",
              "name": "Supplier One",
              "email": "supplier.one@example.com",
              "phone_number": 5551234567,
              "created_at": "2024-03-01T08:00:00Z"
            },
            "products": {
              "id": "f2b2d5eb-76d4-4b09-a64d-9c2fd6a4d19e",
              "name": "Product A",
              "description": "Description of Product A",
              "created_at": "2024-02-01T09:00:00Z"
            }
          }
        }
      ]
    },
    {
      "id": "c3f7d832-04e1-4b77-b19b-5a7c36be83d2",
      "client_id": "c3f7d832-04e1-4b77-b19b-5a7c36be83d1",
      "order_date": "2024-04-05T15:30:00Z",
      "status": "SHIPPED",
      "created_at": "2024-04-05T15:30:00Z",
      "clients": {
        "id": "c3f7d832-04e1-4b77-b19b-5a7c36be83d1",
        "name": "Bob Johnson",
        "email": "bob.johnson@example.com",
        "phone_number": 9876543210,
        "created_at": "2024-01-05T15:30:00Z"
      },
      "order_items": [
        {
          "id": "d8e10f0c-bb27-4d8d-bff6-46f9e925a5b2",
          "order_id": "c3f7d832-04e1-4b77-b19b-5a7c36be83d2",
          "inventory_id": "ac15d5a6-14f1-4a7f-98a3-0e4f2e1c1d4f",
          "quantity": 1,
          "price": 75.00,
          "created_at": "2024-04-05T15:30:00Z",
          "inventory": {
            "id": "ac15d5a6-14f1-4a7f-98a3-0e4f2e1c1d4f",
            "cost": 75.00,
            "supplier_id": "a2b5d5b9-4c3a-4b51-a2c6-123456789abc",
            "product_id": "a4cde3f2-1f98-4b53-b8ee-8f7d64a84b9d",
            "stock_quantity": 50,
            "reorder_limit": 10,
            "supply_date": "2024-03-20T11:00:00Z",
            "created_at": "2024-03-20T11:00:00Z",
            "suppliers": {
              "id": "a2b5d5b9-4c3a-4b51-a2c6-123456789abc",
              "name": "Supplier Two",
              "email": "supplier.two@example.com",
              "phone_number": 5559876543,
              "created_at": "2024-03-10T14:00:00Z"
            },
            "products": {
              "id": "a4cde3f2-1f98-4b53-b8ee-8f7d64a84b9d",
              "name": "Product B",
              "description": "Description of Product B",
              "created_at": "2024-02-15T10:00:00Z"
            }
          }
        }
      ]
    }
  ],
  "invoices": [
    {
      "id": "c7f9d0d5-9a30-4a26-b0c4-00ff18b8b781",
      "order_id": "e4b6a0f1-d09e-4e95-8d1a-8e091b2a7511",
      "amount": 200.00,
      "invoice_date": "2024-04-02T12:00:00Z",
      "order": {
        "id": "e4b6a0f1-d09e-4e95-8d1a-8e091b2a7511",
        "client_id": "e4b6a0f1-d09e-4e95-8d1a-8e091b2a7510",
        "order_date": "2024-04-01T12:00:00Z",
        "status": "PROCESSING",
        "created_at": "2024-04-01T12:00:00Z",
        "clients": {
          "id": "e4b6a0f1-d09e-4e95-8d1a-8e091b2a7510",
          "name": "Alice Smith",
          "email": "alice.smith@example.com",
          "phone_number": 1234567890,
          "created_at": "2024-01-01T12:00:00Z"
        },
        "order_items": [
          {
            "id": "d8e10f0c-bb27-4d8d-bff6-46f9e925a5b1",
            "order_id": "e4b6a0f1-d09e-4e95-8d1a-8e091b2a7511",
            "inventory_id": "edc1b083-5c7f-4d7f-8d52-1c5302a3d7b1",
            "quantity": 2,
            "price": 100.00,
            "created_at": "2024-04-01T12:00:00Z",
            "inventory": {
              "id": "edc1b083-5c7f-4d7f-8d52-1c5302a3d7b1",
              "cost": 50.00,
              "supplier_id": "b9d6d220-4f6c-42d8-9083-b0a5c3f05a9a",
              "product_id": "f2b2d5eb-76d4-4b09-a64d-9c2fd6a4d19e",
              "stock_quantity": 100,
              "reorder_limit": 20,
              "supply_date": "2024-03-15T10:00:00Z",
              "created_at": "2024-03-15T10:00:00Z",
              "suppliers": {
                "id": "b9d6d220-4f6c-42d8-9083-b0a5c3f05a9a",
                "name": "Supplier One",
                "email": "supplier.one@example.com",
                "phone_number": 5551234567,
                "created_at": "2024-03-01T08:00:00Z"
              },
              "products": {
                "id": "f2b2d5eb-76d4-4b09-a64d-9c2fd6a4d19e",
                "name": "Product A",
                "description": "Description of Product A",
                "created_at": "2024-02-01T09:00:00Z"
              }
            }
          }
        ]
      }
    },
    {
      "id": "a9f5d0d4-90ff-4a69-b0b5-7b8e3c5fbb9d",
      "order_id": "c3f7d832-04e1-4b77-b19b-5a7c36be83d2",
      "amount": 75.00,
      "invoice_date": "2024-04-06T15:30:00Z",
      "order": {
        "id": "c3f7d832-04e1-4b77-b19b-5a7c36be83d2",
        "client_id": "c3f7d832-04e1-4b77-b19b-5a7c36be83d1",
        "order_date": "2024-04-05T15:30:00Z",
        "status": "SHIPPED",
        "created_at": "2024-04-05T15:30:00Z",
        "clients": {
          "id": "c3f7d832-04e1-4b77-b19b-5a7c36be83d1",
          "name": "Bob Johnson",
          "email": "bob.johnson@example.com",
          "phone_number": 9876543210,
          "created_at": "2024-01-05T15:30:00Z"
        },
        "order_items": [
          {
            "id": "d8e10f0c-bb27-4d8d-bff6-46f9e925a5b2",
            "order_id": "c3f7d832-04e1-4b77-b19b-5a7c36be83d2",
            "inventory_id": "ac15d5a6-14f1-4a7f-98a3-0e4f2e1c1d4f",
            "quantity": 1,
            "price": 75.00,
            "created_at": "2024-04-05T15:30:00Z",
            "inventory": {
              "id": "ac15d5a6-14f1-4a7f-98a3-0e4f2e1c1d4f",
              "cost": 75.00,
              "supplier_id": "a2b5d5b9-4c3a-4b51-a2c6-123456789abc",
              "product_id": "a4cde3f2-1f98-4b53-b8ee-8f7d64a84b9d",
              "stock_quantity": 50,
              "reorder_limit": 10,
              "supply_date": "2024-03-20T11:00:00Z",
              "created_at": "2024-03-20T11:00:00Z",
              "suppliers": {
                "id": "a2b5d5b9-4c3a-4b51-a2c6-123456789abc",
                "name": "Supplier Two",
                "email": "supplier.two@example.com",
                "phone_number": 5559876543,
                "created_at": "2024-03-10T14:00:00Z"
              },
              "products": {
                "id": "a4cde3f2-1f98-4b53-b8ee-8f7d64a84b9d",
                "name": "Product B",
                "description": "Description of Product B",
                "created_at": "2024-02-15T10:00:00Z"
              }
            }
          }
        ]
      }
    }
  ]
}
