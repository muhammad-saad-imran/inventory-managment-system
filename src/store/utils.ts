import { completeLoading, startLoading } from "@/store/features/loading";
import { createAsyncThunk, GetThunkAPI } from "@reduxjs/toolkit";

export const getThunkCreator = <T>(
  thunkName: string,
  getEntity: (id: string) => Promise<T>
) =>
  createAsyncThunk(
    thunkName,
    async (id: string, { dispatch, rejectWithValue }) => {
      try {
        dispatch(startLoading());
        const entity = await getEntity(id);
        dispatch(completeLoading());
        return entity;
      } catch (error) {
        dispatch(completeLoading());
        alert("Error fetching products");
        rejectWithValue(error);
      }
    }
  );

export const getAllThunkCreator = <T>(
  thunkName: string,
  getAllEntity: (input: string) => Promise<T[]>
) =>
  createAsyncThunk(
    thunkName,
    async (name: string, { dispatch, rejectWithValue }) => {
      try {
        dispatch(startLoading());
        const entity = await getAllEntity(name);
        dispatch(completeLoading());
        return entity;
      } catch (error) {
        dispatch(completeLoading());
        alert("Error fetching products");
        rejectWithValue(error);
      }
    }
  );

export const mutationThunkCreator = <T>(
  thunkName: string,
  mutateEntity: (values: T) => Promise<T>
) =>
  createAsyncThunk(
    thunkName,
    async (values: T, { dispatch, rejectWithValue }) => {
      try {
        dispatch(startLoading());
        const entity = await mutateEntity(values);
        dispatch(completeLoading());
        return entity;
      } catch (error) {
        dispatch(completeLoading());
        alert("Error fetching products");
        rejectWithValue(error);
      }
    }
  );

export const deleteThunkCreator = (
  thunkName: string,
  deleteEntity: (id: string) => Promise<void>
) =>
  createAsyncThunk(
    thunkName,
    async (id: string, { dispatch, rejectWithValue }) => {
      try {
        dispatch(startLoading());
        await deleteEntity(id);
        dispatch(completeLoading());
        return;
      } catch (error) {
        dispatch(completeLoading());
        alert("Error fetching products");
        rejectWithValue(error);
      }
    }
  );
