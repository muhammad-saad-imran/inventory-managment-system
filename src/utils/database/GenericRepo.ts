import { DB_TABLES } from "@/utils/database/types";
import { SupabaseClient } from "@supabase/supabase-js";

export abstract class GenericRepo<T> {
  supabase: SupabaseClient;
  tableName: DB_TABLES;

  constructor(_client: SupabaseClient, _tableName: DB_TABLES) {
    this.supabase = _client;
    this.tableName = _tableName;
  }

  async getAll(include?: string): Promise<T[]> {
    const { data: entity, error } = await this.supabase
      .from(this.tableName)
      .select(include);

    if (entity) {
      return entity as T[];
    } else {
      throw error;
    }
  }

  async get(id: string, include?: string): Promise<T> {
    const { data: entity, error } = await this.supabase
      .from(this.tableName)
      .select(include)
      .eq("id", id)
      // .limit(1)
      .single();

    if (entity) {
      return entity as T;
    } else {
      throw error;
    }
  }

  async create(insertValue: T, include?: string): Promise<T> {
    const { data: entity, error } = await this.supabase
      .from(this.tableName)
      .insert(insertValue)
      .select(include)
      // .limit(1)
      .single();

    if (entity) {
      return entity as T;
    } else {
      throw error;
    }
  }

  async update(id: string, updateValue: T, include?: string): Promise<T> {
    const { data: entity, error } = await this.supabase
      .from(this.tableName)
      .update(updateValue)
      .eq("id", id)
      .select(include)
      // .limit(1)
      .single();

    if (entity) {
      return entity as T;
    } else {
      throw error;
    }
  }

  async delete(id: string, include?: string): Promise<void> {
    const { error } = await this.supabase
      .from(this.tableName)
      .delete()
      .eq("id", id);

    if (error) {
      throw error;
    }
  }
}
