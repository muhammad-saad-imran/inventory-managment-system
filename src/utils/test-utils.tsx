import { PropsWithChildren } from 'react'
import { Provider } from 'react-redux'
import { render, type RenderOptions } from '@testing-library/react'

import { AppStore, makeStore } from '@/store'

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  store?: AppStore
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    // Automatically create a store instance if no store was passed in
    store = makeStore(),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return <Provider store={store}>{children}</Provider>
  }  
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) }
}