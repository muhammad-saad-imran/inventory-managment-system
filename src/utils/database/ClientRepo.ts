import { GenericRepo } from "@/utils/database/GenericRepo";
import { Client, DB_TABLES } from "@/utils/database/types";
import { SupabaseClient } from "@supabase/supabase-js";

export class ClientRepo extends GenericRepo<Client> {
  constructor(_client: SupabaseClient) {
    super(_client, DB_TABLES.CLIENTS);
  }

  async getWithName(name: string): Promise<Client[]> {
    const { data: entity, error } = await this.supabase
      .from(this.tableName)
      .select()
      .ilike("name", `%${name}%`);

    if (entity) {
      return entity;
    } else {
      throw error;
    }
  }
}
