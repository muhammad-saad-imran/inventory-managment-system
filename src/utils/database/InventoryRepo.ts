import { SupabaseClient } from "@supabase/supabase-js";
import { GenericRepo } from "@/utils/database/GenericRepo";
import { DB_TABLES, Inventory } from "@/utils/database/types";

export class InventoryRepo extends GenericRepo<Inventory> {
  constructor(_client: SupabaseClient) {
    super(_client, DB_TABLES.INVENTORY);
  }

  async getWithProductAndSupplier(
    product_id: string,
    supplier_id: string
  ): Promise<Inventory> {
    const { data: entity, error } = await this.supabase
      .from(this.tableName)
      .select("*, suppliers!inner (*), products!inner (*)")
      .eq("products.id", product_id)
      .eq("suppliers.id", supplier_id)
      .limit(1)
      .single();

    if (entity) {
      return entity;
    } else {
      throw error;
    }
  }

  async getWithProductName(
    name: string,
    supplier_id?: string
  ): Promise<Inventory[]> {
    let query = this.supabase
      .from(this.tableName)
      .select("*, suppliers!inner (*), products!inner (*)")
      .ilike("products.name", `%${name}%`)
      

    if (supplier_id) {
      query = query.eq("supplier_id", supplier_id);
    }

    const { data: entity, error } = await query;

    if (entity) {
      return entity;
    } else {
      throw error;
    }
  }

  async getSupplierWithName(
    name: string,
    product_id?: string
  ): Promise<Inventory[]> {
    let query = this.supabase
      .from(this.tableName)
      .select("*, suppliers!inner (*), products!inner (*)")
      .ilike("suppliers.name", `%${name}%`);

    if (product_id) {
      query = query.eq("product_id", product_id);
    }

    const { data: entity, error } = await query;

    if (entity) {
      return entity;
    } else {
      throw error;
    }
  }
}
