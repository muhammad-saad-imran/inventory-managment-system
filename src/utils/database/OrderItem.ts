import { SupabaseClient } from "@supabase/supabase-js";
import { GenericRepo } from "@/utils/database/GenericRepo";
import { DB_TABLES, OrderItem } from "@/utils/database/types";

export class OrderItemRepo extends GenericRepo<OrderItem> {
  constructor(_client: SupabaseClient) {
    super(_client, DB_TABLES.ORDER_ITEMS);
  }
}
