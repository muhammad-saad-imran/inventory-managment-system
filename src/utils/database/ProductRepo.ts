import { GenericRepo } from "@/utils/database/GenericRepo";
import { DB_TABLES, Product } from "@/utils/database/types";
import { SupabaseClient } from "@supabase/supabase-js";

export class ProductRepo extends GenericRepo<Product> {
  constructor(_client: SupabaseClient) {
    super(_client, DB_TABLES.PRODUCTS);
  }

  async getWithName(name: string, include?: string): Promise<Product[]> {
    const { data: entity, error } = await this.supabase
      .from(this.tableName)
      .select(include)
      .ilike("name", `%${name}%`);

    if (entity) {
      return entity as unknown as Product[];
    } else {
      throw error;
    }
  }
}
