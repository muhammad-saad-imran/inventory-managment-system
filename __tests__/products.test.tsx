import { fireEvent, screen, waitFor } from "@testing-library/dom";
import { act } from "@testing-library/react";
import { renderWithProviders } from "@/utils/test-utils";
import { ProductRepo } from "@/utils/database/ProductRepo";
import database from "@/../__mocks__/database.json";
import ProductPage from "@/app/dashboard/products/page";
import CreateProductPage from "@/app/dashboard/products/create/page";
import ProductInfoPage from "@/app/dashboard/products/[id]/page";

jest.mock("next/navigation", () => {
  const actualImport = jest.requireActual("next/navigation");
  return {
    ...actualImport,
    usePathname() {
      return "/dashboard/products";
    },
  };
});

jest.mock("@/utils/supabase/client");
jest.mock("@/utils/database/ProductRepo");

describe("Products feature pages", () => {
  let getWithNameMock = jest.spyOn(ProductRepo.prototype, "getWithName");
  let getProductMock = jest.spyOn(ProductRepo.prototype, "get");
  let createProductMock = jest.spyOn(ProductRepo.prototype, "create");
  let updateProductMock = jest.spyOn(ProductRepo.prototype, "update");
  let deleteProductMock = jest.spyOn(ProductRepo.prototype, "delete");

  let useRouter = jest.spyOn(require("next/navigation"), "useRouter");
  let useParams = jest.spyOn(require("next/navigation"), "useParams");

  const routerPushMock = jest.fn();

  beforeEach(() => {
    // mock useRouter implemention
    useRouter.mockImplementation(() => ({ push: routerPushMock }));

    // mock useParams implemetation
    useParams.mockImplementation(() => ({ id: database.products[0].id }));

    // mock ProductRepo method
    getWithNameMock.mockResolvedValue(database.products);
    getProductMock.mockResolvedValue(database.products[0]);
    createProductMock.mockResolvedValue(database.products[0]);
    updateProductMock.mockResolvedValue(database.products[0]);
  });

  afterEach(() => {
    // Cleanup mocks after each test
    useRouter.mockClear();
    getWithNameMock.mockClear();
    getProductMock.mockClear();
    createProductMock.mockClear();
    deleteProductMock.mockClear();
    updateProductMock.mockClear();
  });

  describe("ProductPage - displays all products", () => {
    it("should match snapshot", async () => {
      let container;

      await act(async () => {
        ({ container } = renderWithProviders(<ProductPage />)); // ARRANGE
      });
      expect(container).toMatchSnapshot(); //ASSERT
    });

    it("should route on clicking product table row", async () => {
      renderWithProviders(<ProductPage />); // ARRANGE

      // ACT
      const elems = await screen.findAllByTestId("product-row");
      fireEvent.click(elems[0]);

      expect(routerPushMock).toHaveBeenCalledTimes(1); // ASSERT
    });
  });

  describe("CreateProductPage", () => {
    it("should match snapshot", async () => {
      let container;

      await act(async () => {
        ({ container } = await renderWithProviders(<CreateProductPage />)); // ARRANGE
      });
      expect(container).toMatchSnapshot(); //ASSERT
    });

    it("should create product", async () => {
      await act(() => renderWithProviders(<CreateProductPage />)); // ARRANGE

      await act(async () => {
        // ACT
        const nameInput = await screen.findByLabelText("Name");
        fireEvent.change(nameInput, { target: { value: "test product" } });

        const descInput = await screen.findByLabelText("Description");
        fireEvent.change(descInput, { target: { value: "test description" } });

        const submitBtn = await screen.findByTestId("submit-btn");
        fireEvent.click(submitBtn);
      });

      // ASSERT
      expect(createProductMock).toHaveBeenCalledTimes(1);
      expect(routerPushMock).toHaveBeenCalledTimes(1);
    });
  });

  describe("ProductInfoPage", () => {
    it("should match snapshot", async () => {
      let container;

      await act(async () => {
        ({ container } = await renderWithProviders(<ProductInfoPage />)); // ARRANGE
      });
      expect(container).toMatchSnapshot(); //ASSERT
    });

    it("should delete product", async () => {
      await act(() => renderWithProviders(<ProductInfoPage />)); // ARRANGE

      await act(async () => {
        // ACT
        const deleteBtn = await screen.findByTestId("delete-btn");
        fireEvent.click(deleteBtn);
      });

      // ASSERT
      expect(deleteProductMock).toHaveBeenCalledTimes(1);
      expect(routerPushMock).toHaveBeenCalledTimes(1);
    });

    it("should update product", async () => {
      await act(() => renderWithProviders(<ProductInfoPage />)); // ARRANGE

      await act(async () => {
        // ACT
        const nameInput = await screen.findByLabelText("Name");
        fireEvent.change(nameInput, {
          target: { value: "Changed Test product" },
        });

        const updateBtn = await screen.findByTestId("update-btn");
        fireEvent.click(updateBtn);
      });

      // ASSERT
      expect(updateProductMock).toHaveBeenCalledTimes(1);
    });
  });
});
