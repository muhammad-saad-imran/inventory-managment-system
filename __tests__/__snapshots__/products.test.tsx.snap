// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Products feature pages CreateProductPage should match snapshot 1`] = `
<div>
  <div>
    <p
      class="text-3xl text-center mb-8"
    >
      Create Product
    </p>
    <form
      class="flex flex-col gap-4 bg-white p-3"
    >
      <div
        class="flex flex-col gap-1 w-full"
      >
        <label
          for="name"
        >
          Name
        </label>
        <input
          class="sc-23ee340d-0 sc-23ee340d-1 cwYaeo jqzeAh"
          data-testid="name"
          id="name"
          name="name"
          value=""
        />
      </div>
      <div
        class="flex flex-col gap-1 w-full"
      >
        <label
          for="description"
        >
          Description
        </label>
        <input
          class="sc-23ee340d-0 sc-23ee340d-1 cwYaeo jqzeAh"
          data-testid="description"
          id="description"
          name="description"
          value=""
        />
      </div>
      <button
        class="sc-7d679c72-0 sc-7d679c72-1 zDViG iTQhfb w-fit"
        data-testid="submit-btn"
        type="submit"
      >
        Create
      </button>
    </form>
  </div>
</div>
`;

exports[`Products feature pages ProductInfoPage should match snapshot 1`] = `
<div>
  <div>
    <p
      class="text-3xl text-center mb-8"
    >
      Product
    </p>
    <form
      class="flex flex-col gap-4 bg-white p-3"
    >
      <div
        class="flex flex-col gap-1 w-full"
      >
        <label
          for="name"
        >
          Name
        </label>
        <input
          class="sc-23ee340d-0 sc-23ee340d-1 cwYaeo jqzeAh"
          data-testid="name"
          id="name"
          name="name"
          value="Product A"
        />
      </div>
      <div
        class="flex flex-col gap-1 w-full"
      >
        <label
          for="description"
        >
          Description
        </label>
        <input
          class="sc-23ee340d-0 sc-23ee340d-1 cwYaeo jqzeAh"
          data-testid="description"
          id="description"
          name="description"
          value="Description of Product A"
        />
      </div>
      <div
        class="flex gap-4"
      >
        <button
          class="sc-7d679c72-0 sc-7d679c72-1 gkXiWL iTQhfb w-fit"
          data-testid="update-btn"
          disabled=""
          type="submit"
        >
          Update
        </button>
        <button
          class="sc-7d679c72-0 sc-7d679c72-1 zDViG iTQhfb w-fit"
          data-testid="delete-btn"
          type="button"
        >
          Delete
        </button>
      </div>
    </form>
  </div>
</div>
`;

exports[`Products feature pages ProductPage - displays all products should match snapshot 1`] = `
<div>
  <div>
    <div>
      <p
        class="text-3xl text-center"
      >
        Products
      </p>
      <div
        class="flex gap-8 bg-white w-full p-5 mt-3"
      >
        <input
          class="sc-23ee340d-0 cwYaeo border-0 text-lg"
          placeholder="Search"
          value=""
        />
        <button
          class="sc-7d679c72-0 sc-7d679c72-1 zDViG iTQhfb"
        >
          <p
            class="text-xl"
          >
            +
          </p>
        </button>
      </div>
    </div>
    <table
      class="w-full mt-3 bg-white"
    >
      <thead>
        <tr>
          <th
            class="py-5"
          >
            Name
          </th>
          <th
            class="py-5"
          >
            Description
          </th>
          <th
            class="py-5"
          >
            Created At
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
  </div>
</div>
`;
